File: Go/manager.go
/**
 * @Author: tangchengqin
 * @Date: 2024/9/18 11:47:38
 * @LastEditors: tangchengqin
 * @LastEditTime: 2024/9/18 16:27:20
 * Description:
 * Copyright: Copyright (©)}) 2024 Clarify. All rights reserved.
 */
package event

import (
	"sync"
)

var (
	instance *EventManager
	once  	 sync.Once
)

type EventList []Event

func GetEventManager() *EventManager {
	once.Do(func() {
		instance = &EventManager{
			eventMap: make(map[string]EventList),
			onceEventMap: make(map[string]EventList),
			counter: 1,
		}
	})
	return instance
}

type EventManager struct {
	eventMap map[string]EventList
	onceEventMap map[string]EventList
	counter int
}

// 默认过滤器
func DefaultFilter(...any) bool {
	return true
}

func (slf *EventManager) RegisterEvent(eventName string, callback func(...any), filter func(...any) bool, once bool) int {
	if once {
		return slf.listenOnce(eventName, callback, filter)
	}
	return slf.listen(eventName, callback, filter)
}

func (slf *EventManager) listen(eventName string, callback func(...any), filter func(...any) bool) int {
	eventList, exist := slf.eventMap[eventName]
	if !exist {
		eventList = EventList{}
	}
	evtId := slf.counter
	eventList = append(eventList, Event{evtId: evtId, callback: callback, eventName: eventName, filter: filter})
	slf.eventMap[eventName] = eventList
	slf.counter++
	return evtId
}

func (slf *EventManager) listenOnce(eventName string, callback func(...any), filter func(...any) bool) int {
	eventList, exist := slf.onceEventMap[eventName]
	if !exist {
		eventList = EventList{}
	}
	evtId := slf.counter
	eventList = append(eventList, Event{evtId: evtId, callback: callback, eventName: eventName, filter: filter})
	slf.onceEventMap[eventName] = eventList
	slf.counter++
	return evtId
}

func (slf *EventManager) RemoveEvent(eventName string, evtId interface{}) {
	if evtId == nil {
		slf.removeAll(eventName)
		return
	}
	slf.remove(eventName, evtId.(int))
	slf.removeOnce(eventName, evtId.(int))
}

func (slf *EventManager) remove(eventName string, evtId int) {
	eventList, exist := slf.eventMap[eventName]
	if !exist {
		return
	}
	for i, event := range eventList {
		if event.GetID() != evtId {
			continue
		}
		eventList = append(eventList[:i], eventList[i+1:]...)
		slf.eventMap[eventName] = eventList
		return
	}
}

func (slf *EventManager) removeOnce(eventName string, evtId int) {
	eventList, exist := slf.onceEventMap[eventName]
	if !exist {
		return
	}
	for i, event := range eventList {
		if event.GetID() != evtId {
			continue
		}
		eventList = append(eventList[:i], eventList[i+1:]...)
		slf.onceEventMap[eventName] = eventList
		return
	}
}

func (slf *EventManager) removeAll(eventName string) {
	delete(slf.eventMap, eventName)
	delete(slf.onceEventMap, eventName)
}

func (slf *EventManager) TriggerEvent(eventName string, evtId int, args ...any) {
	slf.execute(eventName, evtId, args...)
	slf.executeOnce(eventName, evtId, args...)
}

func (slf *EventManager) execute(eventName string, evtId int, args ...any) {
	eventList, exist := slf.eventMap[eventName]
	if !exist {
		return
	}
	if evtId == -1 {
		for _, event := range eventList {
			if event.GetID() != evtId {
				continue
			}
			event.Execute(args...)
			return
		}
	}
	for _, event := range eventList {
		event.Execute(args...)
	}
}

func (slf *EventManager) executeOnce(eventName string, evtId int, args ...any) {
	eventList, exist := slf.onceEventMap[eventName]
	if !exist {
		return
	}
	if evtId == -1 {
		for i, event := range eventList {
			if event.GetID() != evtId {
				continue
			}
			event.Execute(args...)
			eventList = append(eventList[:i], eventList[i+1:]...)
			slf.onceEventMap[eventName] = eventList
			return
		}
	}
	for _, event := range eventList {
		event.Execute(args...)
	}
	delete(slf.onceEventMap, eventName)
}


type EventHandler struct {
	EventManager *EventManager
}

func (slf *EventHandler) OnInit() error {
	slf.EventManager = GetEventManager()
	return nil
}

func (slf *EventHandler) Register(eventName string, callback func(...any), filter func(...any) bool, once bool) int {
	return slf.EventManager.RegisterEvent(eventName, callback, filter, once)
}

func (slf *EventHandler) Deregister(eventName string, evtId interface{}) {
	slf.EventManager.RemoveEvent(eventName, evtId)
}

func (slf *EventHandler) OnEvent(eventName string, args ...any) {
	slf.EventManager.TriggerEvent(eventName, -1, args...)
}

func (slf *EventHandler) OnSEvent(eventName string, evtId int, args ...any) {
	slf.EventManager.TriggerEvent(eventName, evtId, args...)
}

File: TypeScript/manager.ts
/**
 * @Author: tangchengqin
 * @Date: 2024/9/14 10:50:35
 * @LastEditors: tangchengqin
 * @LastEditTime: 2024/9/18 16:08:37
 * Description: 
 * Copyright: Copyright (©)}) 2024 Clarify. All rights reserved.
 */

import { Event, Params } from "./event";

class EventManager {

    private static _instance: EventManager;
    private _eventMap: Map<string, Event[]>;
    private _onceEventMap: Map<string, Event[]>;
    private counter: number = 1;

    constructor() {
        this._eventMap = new Map<string, Event[]>();
        this._onceEventMap = new Map<string, Event[]>();
    }

    static getInstance(): EventManager {
        if (!this._instance) {
            this._instance = new EventManager();
        }
        return this._instance;
    }

    private listen(eventName: string, callback: Function, filter?: Function): number {
        let callbacks = this._eventMap.get(eventName);
        if (!callbacks) {
            callbacks = [];
            this._eventMap.set(eventName, callbacks);
        }
        const evtId = this.counter;
        const event = new Event(evtId, eventName, callback, filter);
        callbacks.push(event);
        this.counter ++;
        return evtId;
    }

    private listenOnce(eventName: string, callback: Function, filter?: Function): number {
        let callbacks = this._onceEventMap.get(eventName);
        if (!callbacks) {
            callbacks = [];
            this._onceEventMap.set(eventName, callbacks);
        }
        const evtId = this.counter;
        const event = new Event(evtId, eventName, callback, filter);
        callbacks.push(event);
        this.counter ++;
        return evtId;
    }

    public on(eventName: string, callback: Function, filter?: Function, once?: boolean): number {
        if (once) {
            return this.listenOnce(eventName, callback, filter);
        }
        return this.listen(eventName, callback, filter);
    }

    private cancel(eventName: string, evtId: number): void {
        const events = this._eventMap.get(eventName);
        if (!events) {
            return;
        }
        for (let i = 0; i < events.length; i++) {
            const event = events[i];
            if (event.evtId !== evtId) {
                continue;
            }
            events.splice(i, 1);
            break;
        }
    }

    private cancelOnce(eventName: string, evtId: number): void {
        const events = this._onceEventMap.get(eventName);
        if (!events) {
            return;
        }
        for (let i = 0; i < events.length; i++) {
            const event = events[i];
            if (event.evtId !== evtId) {
                continue;
            }
            events.splice(i, 1);
            break;
        }
    }

    private cancelAll(eventName: string): void {
        this._eventMap.delete(eventName);
        this._onceEventMap.delete(eventName);
    }

    public off(eventName: string, evtId?: number, once?: boolean) {
        if (!evtId) {
            this.cancelAll(eventName);
            return;
        }
        if (once) {
            this.cancelOnce(eventName, evtId);
            return;
        }
        this.cancel(eventName, evtId);
    }

    private execute(eventName: string, args: Params, evtId?: number): void {
        const events = this._eventMap.get(eventName);
        if (!events) {
            return;
        }
        if (evtId) {
            for (const event of events) {
                if (event.evtId === evtId) {
                    event.execute(args);
                    return;
                }
            }
        } else {
            for (const event of events) {
                event.execute(args);
            }
        }
    }

    private executeOnce(eventName: string, args: Params, evtId?: number): void {
        const events = this._onceEventMap.get(eventName);
        if (!events) {
            return;
        }
        if (evtId) {
            for (const event of events) {
                if (event.evtId === evtId) {
                    event.execute(args);
                    events.splice(events.indexOf(event), 1);
                    return;
                }
            }
        } else {
            for (const event of events) {
                event.execute(args);
            }
            this._onceEventMap.delete(eventName);
        }
    }

    public call(eventName: string, args: Params, evtId?: number): void {
        this.execute(eventName, args, evtId);
        this.executeOnce(eventName, args, evtId);
    }

}

const handler = {

    listen(eventName: string, callback: Function, filter?: Function, once?: boolean): number {
        return EventManager.getInstance().on(eventName, callback, filter, once);
    },

    listenMulti(eventList: Array<any>): Array<number> {
        const ret: Array<number> = [];
        for (const event of eventList) {
            const evtId = EventManager.getInstance().on(event.eventName, event.callback, event.filter, event.once);
            ret.push(evtId);
        }
        return ret;
    },

    cancel(eventName: string, evtId?: number, once?: boolean) {
        EventManager.getInstance().off(eventName, evtId, once);
    },

    onEvent(eventName: string, args: any, evtId?: number) {
        EventManager.getInstance().call(eventName, args, evtId);
    }

}

export const installEventSystem = (target: any) => {
    for (const key in handler) {
        const func = handler[key];
        target[key] = func.bind(target);
    }
}
